### ISSUE TITLE
Initialization troubles with teacup 1.2.9 and ruby 2.0.0

### ISSUE THREAD

**Issue by robheittman** [colinta/teacup#69]
Updating to teacup 1.2.9 on RubyMotion 1.35, it's compiling and requiring but failing to initialize, even after "rake clean", simulator reset, clean checkout and build, etc. Downgrading to 1.2.8 gets it working again. Other gems in the project (bubble-wrap, motion-settings-bundle, motion-cocoapods) aren't having trouble and have been updated OK. Poked around for a bit but couldn't make sense of it. No special detect_dependencies settings in my Rakefile ... it's just doing whatever it does.

I couldn't come up with any good hypothesis as to why, even after reading the commits between 1.2.8 and 1.2.9.

```
2013-04-07 14:35:00.272 NFC Actions[66058:c07] uninitialized constant Teacup::Layout (NameError)
2013-04-07 14:35:00.275 NFC Actions[66058:c07] *** Terminating app due to uncaught exception 'NameError', reason: 'uninitialized constant Teacup::Layout (NameError)
'
*** First throw call stack:
(0x2864012 0x4bbe7e 0x232c14 0xe3f2 0xdc2d)
libc++abi.dylib: terminate called throwing an exception
((null))> *** simulator session ended with error: Error Domain=DTiPhoneSimulatorErrorDomain Code=1 "The simulated application quit." UserInfo=0x10012b720 {NSLocalizedDescription=The simulated application quit., DTiPhoneSimulatorUnderlyingErrorCodeKey=-1}
rake aborted!
Command failed with status (1): [DYLD_FRAMEWORK_PATH="/Applications/Xcode.a...]
```

**Comment by colinta** [comment]
I'll try to reproduce it - can you try and narrow down what file it is?  Just putting `NSLog` at the top of a few files might do the trick...

**Comment by robheittman** [comment]
Looks like lib/teacup/z_core_extensions/ui_view_controller.rb:4

Augmenting all the classes with NSLogs yields this:

```
2013-04-07 15:27:02.357 Issue69[71894:c07] begin lib/teacup/calculations.rb
2013-04-07 15:27:02.364 Issue69[71894:c07] end lib/teacup/calculations.rb
2013-04-07 15:27:02.366 Issue69[71894:c07] begin lib/teacup/stylesheet_extensions/calculations.rb
2013-04-07 15:27:02.369 Issue69[71894:c07] end lib/teacup/stylesheet_extensions/calculations.rb
2013-04-07 15:27:02.370 Issue69[71894:c07] begin lib/teacup/z_core_extensions/ui_view_controller.rb
2013-04-07 15:27:02.475 Issue69[71894:c07] uninitialized constant Teacup::Layout (NameError)
2013-04-07 15:27:02.476 Issue69[71894:c07] *** Terminating app due to uncaught exception 'NameError', reason: 'uninitialized constant Teacup::Layout (NameError)
'
```

No idea why ... looks like lib/teacup/layout.rb was compiled:

```
   Compile /Users/robheittman/.rvm/gems/ruby-2.0.0-p0@issue69/gems/teacup-1.2.9/lib/teacup/layout.rb
```

**Comment by colinta** [comment]
I wonder if ruby 2.0 has anything to do with it - doubtful, but i noticed that, and i'm using 1.9.3.

for me, the order is very different:

```
2013-04-07 13:38:31.657 teacup-test[2021:11303] calculations.rb
2013-04-07 13:38:31.662 teacup-test[2021:11303] version.rb
2013-04-07 13:38:31.663 teacup-test[2021:11303] constraint.rb
2013-04-07 13:38:31.676 teacup-test[2021:11303] handler.rb
2013-04-07 13:38:31.689 teacup-test[2021:11303] rotation.rb
2013-04-07 13:38:31.692 teacup-test[2021:11303] layout.rb
2013-04-07 13:38:31.695 teacup-test[2021:11303] merge_defaults.rb
2013-04-07 13:38:31.707 teacup-test[2021:11303] restyle.rb
2013-04-07 13:38:31.708 teacup-test[2021:11303] style.rb
2013-04-07 13:38:31.732 teacup-test[2021:11303] stylesheet.rb
2013-04-07 13:38:31.734 teacup-test[2021:11303] autoresize.rb
2013-04-07 13:38:31.737 teacup-test[2021:11303] constraints.rb
2013-04-07 13:38:31.743 teacup-test[2021:11303] geometry.rb
2013-04-07 13:38:31.748 teacup-test[2021:11303] ui_view_getters.rb
2013-04-07 13:38:31.759 teacup-test[2021:11303] teacup_util.rb
2013-04-07 13:38:31.763 teacup-test[2021:11303] ca_layer.rb
2013-04-07 13:38:31.770 teacup-test[2021:11303] ui_view.rb
2013-04-07 13:38:31.772 teacup-test[2021:11303] ui_view_controller.rb
2013-04-07 13:38:31.775 teacup-test[2021:11303] z_handlers.rb
2013-04-07 13:38:31.779 teacup-test[2021:11303] dummy.rb
```

the compile order is (these could be different, but they should be similar at least)

   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/calculations.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/version.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/handler.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/constraint.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/layout.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/stylesheet_extensions/rotation.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/restyle.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/merge_defaults.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/stylesheet.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/style.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/stylesheet_extensions/constraints.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/stylesheet_extensions/autoresize.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/z_core_extensions/ui_view_getters.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/stylesheet_extensions/geometry.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/z_core_extensions/ca_layer.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/teacup_util.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/z_core_extensions/ui_view_controller.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/z_core_extensions/ui_view.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/dummy.rb
   Compile /Users/colinta/.rvm/gems/ruby-1.9.3-p194@foo/gems/teacup-1.2.9/lib/teacup/z_core_extensions/z_handlers.rb

(i'm installing ruby 2.0 to try out that theory)

**Comment by robheittman** [comment]
Confirmed 1.9.3 works for me ... seems like a Ruby 2.0.0 issue.

**Comment by colinta** [comment]
Thanks for trying that - 2.0 is still installing on my system... i'll put a heads up in the README and on the google-group mailing list.

**Comment by colinta** [comment]
Actually, gonna leave this open until I figure out why 2.0.0 doesn't work.

**Comment by kastiglione** [comment]
The early evidence points to a difference in the dependency detection code when run on 1.9 vs 2.0. This is a diff of the compile order on 1.9 vs 2.0.

``` diff
diff --git a/1.9.txt b/2.0.txt
index bbf581d..7dbaf2b 100644
--- a/1.9.txt
+++ b/2.0.txt
@@ -1,20 +1,20 @@
 teacup/calculations.rb
-teacup/version.rb
+teacup/stylesheet_extensions/rotation.rb
+teacup/z_core_extensions/ui_view_controller.rb
 teacup/constraint.rb
 teacup/handler.rb
-teacup/stylesheet_extensions/rotation.rb
 teacup/layout.rb
 teacup/merge_defaults.rb
 teacup/restyle.rb
 teacup/style.rb
 teacup/stylesheet.rb
 teacup/stylesheet_extensions/autoresize.rb
 teacup/stylesheet_extensions/constraints.rb
 teacup/stylesheet_extensions/geometry.rb
 teacup/z_core_extensions/ui_view_getters.rb
 teacup/teacup_util.rb
+teacup/version.rb
 teacup/z_core_extensions/ca_layer.rb
 teacup/z_core_extensions/ui_view.rb
-teacup/z_core_extensions/ui_view_controller.rb
 teacup/z_core_extensions/z_handlers.rb
 dummy.rb
```

**Comment by colinta** [comment]
ahh.  hmm...

having ui_view_controller.rb before layout.rb IS the culprit.  thanks, Dave!

#colinta

**Comment by kastiglione** [comment]
EDIT: Below is just a theory, I may have gotten something wrong…

I'm haven't narrowed down the difference between 1.9 and 2.0, but I can say that it looks like [`ce130c1`](https://github.com/rubymotion/teacup/commit/ce130c1e8c7ab3a4196ccf9a799a47b39fc2ef6d#L0R233) is the change that triggered the chain reaction.

By opening up `Teacup` to add `get_subviews`, RubyMotion's dependency detection considers `ui_view_controller.rb` to "define" the `Teacup` module, even though it is defined elsewhere as well. For the purposes of dependency detection, RubyMotion allows many files do "define" a give module / class.

Long story short, this introduced a dependency cycle when run on 2.0 but not 1.9, and RubyMotion's dependency detection doesn't attempt to do anything smart about dependency cycles, it's essentially first come first served. A possible work-around is to move `Teacup.get_subviews` to another file, any file that already opens the `Teacup` module.

**Comment by kastiglione** [comment]
It looks like it's a [`Ripper`](http://www.ruby-doc.org/stdlib-1.9.3/libdoc/ripper/rdoc/Ripper.html) bug in Ruby 1.9. 

With Ruby 2.0, `Ripper::SexpBuilder` calls the `on_const_ref` callback when `ui_view_controller.rb` opens `Teacup`, which seems like the correct behavior. On 1.9, the callback is not called.

**Comment by robheittman** [comment]
Some amazing detective work there, Dave! Confirming I was able to move Teacup.get_subviews to lib/calculations.rb (making no sense, but I picked it because it was first in my 2.0 load order anyway) and fix both my test case and my real application on 2.0.

**Comment by colinta** [comment]
Craaaaaaazy.

So, what's the "good fix" for this?

#colinta

**Comment by kastiglione** [comment]
The real answer is RubyMotion's dependency detection needs to become a little more sophisticated.

I'm away from my computer, but to keep it in ui_view_controller.rb, you could try:

```
def Teacup.get_subviews(target)
  # ...
end
```

EDIT: Without nesting in `module Teacup`.

**Comment by colinta** [comment]
I merged Dave's fix into 1.3.1.

Thanks!

**Comment by kastiglione** [comment]
I don't know why I didn't think of this earlier, but the best fix is to use @clayallsopp's [motion-require](https://github.com/clayallsopp/motion-require). He's converted [formotion](https://github.com/clayallsopp/formotion) to use for managing dependencies.
